core/preconditioner/gauss_seidel.cpp

```diff
@@ -0,0 +1,78 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/preconditioner/gauss_seidel.hpp>
+#include <ginkgo/core/preconditioner/sor.hpp>
+
+#include "core/config/config_helper.hpp"
+
+
+namespace gko {
+namespace preconditioner {
+
+
+template <typename ValueType, typename IndexType>
+typename GaussSeidel<ValueType, IndexType>::parameters_type
+GaussSeidel<ValueType, IndexType>::parse(
+    const config::pnode& config, const config::registry& context,
+    const config::type_descriptor& td_for_child)
+{
+    auto params = GaussSeidel::build();
+
+    if (auto& obj = config.get("skip_sorting")) {
+        params.with_skip_sorting(config::get_value<bool>(obj));
+    }
+    if (auto& obj = config.get("symmetric")) {
+        params.with_symmetric(config::get_value<bool>(obj));
+    }
+    if (auto& obj = config.get("l_solver")) {
+        params.with_l_solver(
+            gko::config::parse_or_get_factory<const LinOpFactory>(
+                obj, context, td_for_child));
+    }
+    if (auto& obj = config.get("u_solver")) {
+        params.with_u_solver(
+            gko::config::parse_or_get_factory<const LinOpFactory>(
+                obj, context, td_for_child));
+    }
+
+    return params;
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<typename GaussSeidel<ValueType, IndexType>::composition_type>
+GaussSeidel<ValueType, IndexType>::generate(
+    std::shared_ptr<const LinOp> system_matrix) const
+{
+    auto product =
+        std::unique_ptr<composition_type>(static_cast<composition_type*>(
+            this->LinOpFactory::generate(std::move(system_matrix)).release()));
```

Comment:
and use generate_impl directly?

Reply:
the `static_cast` is necessary, because `LinOpFactory::generate` returns a `LinOp*`, which can't be used in the constructor of the unique_ptr. The `make_unique` also doesn't work, because the `composition_type` constructor doesn't take a `LinOp*`.

`generate_impl` is not used directly, because this could require to copy-paste the implementation of `LinOpFactory::generate` here, so that it can log correctly and copy to the right executor.